File: Matlab/toolbox/fuzzy/fuzzy/src/defuzz.c
/* Copyright 1994-2002 The MathWorks, Inc.  */
/* $Revision: $  $Date: $  */

/***********************************************************************
 Defuzzification methods
 **********************************************************************/

/* return the center of area of combined output MF (specified by mf)
   of output m */
/* numofpoints is the number of partition for integration */
static DOUBLE defuzzCentroid(FIS *fis, int m, DOUBLE *mf, int numofpoints)
{
	DOUBLE min = fis->output[m]->bound[0];
	DOUBLE max = fis->output[m]->bound[1];
	DOUBLE step = (max - min)/(numofpoints - 1);
	DOUBLE total_mf = 0;
	DOUBLE sum = 0;
	int i;

	for (i = 0; i < numofpoints; i++){
		total_mf += mf[i];
       		sum += mf[i]*(min + step*i);
	}
	if (total_mf == 0) {
		PRINTF("Total area is zero in defuzzCentroid() for output %d.\n", m+1);
		PRINTF("Average of the range of this output variable is used as the output value.\n\n");
		return((fis->output[m]->bound[0] + fis->output[m]->bound[1])/2);
	} 
	return(sum/total_mf);
}

/* return the bisector of area of mf */
static DOUBLE defuzzBisector(FIS *fis, int m, DOUBLE *mf, int numofpoints)
{
	DOUBLE min = fis->output[m]->bound[0];
	DOUBLE max = fis->output[m]->bound[1];
	DOUBLE step = (max - min)/(numofpoints - 1); 
	DOUBLE area, sub_area;
	int i;

	/* find the total area */
	area = 0;
	for (i = 0; i < numofpoints; i++)
		area += mf[i];

	if (area == 0) {
		PRINTF("Total area is zero in defuzzBisector() for output %d.\n", m+1);
		PRINTF("Average of the range of this output variable is used as the output value.\n");
		return((fis->output[m]->bound[0] + fis->output[m]->bound[1])/2);
	} 
     
	sub_area = 0;
	for (i = 0; i < numofpoints; i++) {
		sub_area += mf[i];
		if (sub_area >= area/2)
			break;
	}
	return(min + step*i);
}

/* Returns the mean of maximizing x of mf */
static DOUBLE defuzzMeanOfMax(FIS *fis, int m, DOUBLE *mf, int numofpoints)
{
	DOUBLE min = fis->output[m]->bound[0];
	DOUBLE max = fis->output[m]->bound[1];
	DOUBLE step = (max - min)/(numofpoints - 1); 
	DOUBLE mf_max;
	DOUBLE sum;
	int count;
	int i;

	mf_max = fisArrayOperation(mf, numofpoints, fisMax);

	sum = 0;
	count = 0;
	for (i = 0; i < numofpoints; i++)
		if (mf[i] == mf_max) {
			count++;
			sum += i;
		}
	return(min+step*sum/count);
}

/* Returns the smallest (in magnitude) maximizing x of mf */
static DOUBLE defuzzSmallestOfMax(FIS *fis, int m, DOUBLE *mf, int numofpoints)
{
	DOUBLE min = fis->output[m]->bound[0];
	DOUBLE max = fis->output[m]->bound[1];
	DOUBLE step = (max - min)/(numofpoints - 1); 
	DOUBLE mf_max;
	int i, min_index = 0;
	DOUBLE min_distance = pow(2.0, 31.0)-1;
	DOUBLE distance; /* distance to the origin */

	mf_max = fisArrayOperation(mf, numofpoints, fisMax);
	for (i = 0; i < numofpoints; i++)
		if (mf[i] == mf_max) {
			distance = ABS(min + step*i);
			if (min_distance > distance) {
				min_distance = distance;
				min_index = i;
			}
		}
	return(min + step*min_index);
}

/* Returns the largest (in magnitude) maximizing x of mf */
static DOUBLE defuzzLargestOfMax(FIS *fis, int m, DOUBLE *mf, int numofpoints)
{
	DOUBLE min = fis->output[m]->bound[0];
	DOUBLE max = fis->output[m]->bound[1];
	DOUBLE step = (max - min)/(numofpoints - 1); 
	DOUBLE mf_max;
	int i, max_index = 0;
	DOUBLE max_distance = -(pow(2.0, 31.0)-1);
	DOUBLE distance; /* distance to the origin */

	mf_max = fisArrayOperation(mf, numofpoints, fisMax);
	for (i = 0; i < numofpoints; i++)
		if (mf[i] == mf_max) {
			distance = ABS(min + step*i);
			if (max_distance < distance) {
				max_distance = distance;
				max_index = i;
			}
		}
	return(min + step*max_index);
}

File: Matlab/toolbox/fuzzy/fuzzy/defuzz.m
function out = defuzz(x, mf, type)
% DEFUZZ Defuzzify membership function.
%   Synopsis
%   out = defuzz(x,mf,type)
%   Description
%   defuzz(x,mf,type) returns a defuzzified value out, of a membership function
%   mf positioned at associated variable value x, using one of several
%   defuzzification strategies, according to the argument, type. The variable
%   type can be one of the following.
%   centroid: centroid of area method.
%   bisector: bisector of area method.
%   mom: mean of maximum method.
%   som: smallest of maximum method.
%   lom: largest of maximum method.
%   If type is not one of the above, it is assumed to be a user-defined
%   function. x and mf are passed to this function to generate the defuzzified
%   output.
%   Examples
%   x = -10:0.1:10;
%   mf = trapmf(x,[-10 -8 -4 7]);
%   xx = defuzz(x,mf,'centroid');
%
%   Try DEFUZZDM for more examples.

%   Roger Jang, 6-28-93 ,10-5-93, 9-29-94.
%   Copyright 1994-2002 The MathWorks, Inc. 
%   $Revision: 1.12 $  $Date: 2002/04/14 22:20:23 $

x = x(:);
mf = mf(:);
if length(x) ~= length(mf),
    error('Sizes mismatch!');
end
data_n = length(x);
 
if strcmp(type, 'centroid'),
    total_area = sum(mf);
    if total_area == 0,
        error('Total area is zero in centroid defuzzification!');
    end
    out = sum(mf.*x)/total_area;
    return;
elseif strcmp(type, 'bisector'),
        total_area = sum(mf);
    if total_area == 0,
        error('Total area is zero in bisector defuzzification!');
    end
        tmp = 0;
        for k=1:data_n,
                tmp = tmp + mf(k);
                if tmp >= total_area/2,
                        break;
                end
        end
    out = x(k);
    return;
elseif strcmp(type, 'mom'),
        out = mean(x(find(mf==max(mf))));
    return;
elseif strcmp(type, 'som'),
        tmp = x(find(mf == max(mf)));
        [junk, which] = min(abs(tmp));
    out = tmp(which); 
    return;
elseif strcmp(type, 'lom'),
        tmp = x(find(mf == max(mf)));
        [junk, which] = max(abs(tmp));
    out = tmp(which); 
    return;
else
    % defuzzification type is unknown
    % We assume it is user-defined and evaluate it here
    evalStr=[type '(x, mf)'];
    out = eval(evalStr);
    return;
end
